.github/codeql/queries/token-not-used.ql

```diff
@@ -0,0 +1,41 @@
+/**
+ * @name Don't ignore the token for a cancelable progress bar
+ * @kind problem
+ * @problem.severity warning
+ * @id vscode-codeql/token-not-used
+ * @description If we call `withProgress` with `cancellable: true` but then
+ * ignore the token that is given to us, it will lead to a poor user experience
+ * because the progress bar will appear to be canceled but it will not actually
+ * affect the background process.
+ */
+
+import javascript
+
+class NewTokenSource extends CallExpr {
+  NewTokenSource() {
+    this.getCalleeName() = "withProgress" or this.getCalleeName() = "withInheritedProgress"
+  }
+
+  Function getCallback() {
+    this.getCalleeName() = "withProgress" and result = this.getArgument(0)
+    or
+    this.getCalleeName() = "withInheritedProgress" and result = this.getArgument(1)
```

Comment:
Can we factor this out into a private predicate `isInherited` such that we're not repeating these names 6 times in this file?

Reply:
I think I see what you mean. Sadly that would only reduce it to 4 times because annoyingly I believe you can't reference member predicates in the class predicate thing (I can't remember the real name).

We could make it an abstract class with two subclasses and that way we avoid all duplication. I end using abstract classes in a bunch of my queries so I hope I'm not overusing them, but they are quite useful.

I'll make this change and push it so you can see what I mean.